### PR TITLE
add power module

### PR DESCRIPTION
**Pull request by lee-daly**
Power module which allows an application to modify power attributes of cpu.

Signed-off-by: Lee Daly <lee.daly@intel.com>
Signed-off-by: Anatoly Burakov <anatoly.burakov@intel.com>

### CODE REVIEW

**Comment by lee-daly** [comment]
@anatolyburakov

File: README.md [hunk 1]
```diff
@@ -19,4 +19,7 @@ Intel® Speed Select Technology - Base Frequency (Intel® SST-BF).
 This allows some cores to run at a higher base frequency than
 others.
 
+[pwr.py](pwr.md) module which allows an application to measure/modify core
```
Comment (anatolyburakov) [inline README.md]: i think it is worth stressing that this is not an application but a library.

File: pwr.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
+# pwr module
+
+This is a python module which allows an application to modify power attributes of a CPU. Manipulation can be done to the frequency of cores, 
+frequnecy of the uncore, frequency profiles can be set to achieve desired performance, as well as that capabilities of a CPU can be obtained and its stats monitered.
```
Comment (anatolyburakov) [inline pwr.md]: monitored?

File: pwr.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
+# pwr module
+
+This is a python module which allows an application to modify power attributes of a CPU. Manipulation can be done to the frequency of cores, 
+frequnecy of the uncore, frequency profiles can be set to achieve desired performance, as well as that capabilities of a CPU can be obtained and its stats monitered.  
+Configuration is done by first creating core and/or CPU object lists, these objects will store the characteristics of the physical processor.
+Modifying those objects and committing the changes enables the module to take the necessary steps to make those changes on the CPU hardware.
```
Comment (anatolyburakov) [inline pwr.md]: I don't think you need to go into that many details. Just saying something like "The library will provide a list of core/CPU objects, whose attributes can be changed and committed to the cores/CPUs" or similar.

File: pwr.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+```python
+for c in cores:
+    c.refresh_stats()
+    if c.sst_bf_configured:
+        c.commit("sst_bf_high_turbo")
+```
+
+## Object Referencing
+On object list creation, with `get_cores()/get_cpus()`, *both* object lists are created from *either* call.
```
Comment (anatolyburakov) [inline pwr.md]: No one needs to know these internal details _in a readme_, it doesn't matter when the objects are created and by calling which function. We provide an API which guarantees that the user will get a list of objects. How they got there is an implementation detail and shouldn't be in the readme.

File: pwr.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+This is due to fact that both objects reference each other, the list of core objects can be accessed through the CPU and the CPU object of a core can be accessed through a core.
```
Comment (anatolyburakov) [inline pwr.md]: Similar here: internal details aren't needed. What _does_ matter is getting across that CPU and core objects will reference each other, and it is possible to get CPU from core and vice versa.

File: pwr.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    if core.CPU.SST_BF_ENABLED: #Access CPU flag from core object
+        if core.HIGH_PRIORITY: #Check if core has high priority base frequency 
+            core.min_freq = core.base_freq #Modify minimum desired frequency
+            core.max_freq = core.sst_bf_base_freq #Modify maximum desired frequency
+            core.commit() #Commit changes to hardware
+```
+or
+```python
+for cpu in cpus:
+    if cpu.CORE_LIST[0].epp = "performance" #Access core performance profile through core list of CPU object
```
Comment (anatolyburakov) [inline pwr.md]: Aren't these all lower case now? Also, this line isn't valid Python code.

File: pwr.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    if c.sst_bf_configured:
+        c.commit("sst_bf_high_turbo")
+```
+
+## Object Referencing
+On object list creation, with `get_cores()/get_cpus()`, *both* object lists are created from *either* call.
+This is due to fact that both objects reference each other, the list of core objects can be accessed through the CPU and the CPU object of a core can be accessed through a core.
+
+```python
+for core in cores: #Loop through cores in object list
```
Comment (anatolyburakov) [inline pwr.md]: PEP8 wants two spaces before a comment, and a space after # when writing a comment :)

File: pwr.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+>Specific frequencies will depend on system and configuration.
+
+```python
+For c in cores:
+    c.commit("sst_bf_base") # Configure cores with the SST_BF configuration
+```
+
+## Refreshing cpu stats
+On library initialization the systems capabilities and stats are retrieved.
+CPU stats can become out of date, such as `curr_freq` or `sst_bf_configured`. These can be refreshed in with `refresh_stats()`.
```
Comment (anatolyburakov) [inline pwr.md]: cur?

File: pwr.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+for core in cores:
+    core.min_freq = core.base_freq
+    core.max_freq = core.highest_freq
+    core.commit()
+```
+
+
+### Pre-set Profiles
+When an application is modifying the desired min and max core frequencies, pre-set configurations can also be applied, these will overwrite current configurations and commit the pre-sets.
```
Comment (anatolyburakov) [inline pwr.md]: I would add a note here that the MHz values are based on a specific SKU and can be different for other SKU's.

File: pwr.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+
+### Pre-set Profiles
+When an application is modifying the desired min and max core frequencies, pre-set configurations can also be applied, these will overwrite current configurations and commit the pre-sets.  
+* `minimum`:            Set all cores minimum to 800Mhz and maximum to 800Mhz.
+* `maximum`:            Set all cores minimum to 3900Mhz and maximum to 3900Mhz.
+* `base`:               Set all cores minimum to 2300Mhz and maximum to 2300Mhz.
+* `default`:            Set all cores minimum to 800Mhz and maximum to 3900Mhz.
+* `no_turbo`:           Set all cores minimum to 800Mhz and maximum to 2300Mhz.
+* `sst_bf_base`:        Set high priority cores minimum and maximum to 2700Mhz, normal priority cores minimum and maximum to 2100Mhz.
```
Comment (anatolyburakov) [inline pwr.md]: I would add a note here that these profiles are valid only when SST-BF is enabled. They should probably be in a separate section as "additional profiles only available when SST-BF is enabled".

**Comment by anatolyburakov** [comment]
General comment: this doesn't look like _a library_ which can be downloaded and installed with pip. It looks like a python file that everyone should download and make it part of their own code, and import it from local directory.

I think it would've been worth creating an actual, installable library out of this.

File: pwr.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
+# pwr module
+
+This is a python module which allows an application to modify power attributes of a CPU. Manipulation can be done to the frequency of cores,
```
Comment (anatolyburakov) [inline pwr.md]: I initially wanted to suggest using doxygen to document the library, but Doxygen actually poorly supports Python code and doesn't have a native way to document object attributes (only methods). So, disregard this :D

File: pwr.md [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+* `sst_bf_base_freq`        # priority based frequency [2100Mhz/2700Mhz]
+* `all_core_turbo_freq`     # all core turbo frequency [2800Mhz]
+* `highest_freq`            # highest frequency available [3900Mhz]
+* `lowest_freq`             # lowest active frequency [800Mhz]
+* `curr_freq`               # current core frequency
+* `min_freq`                # desired low frequency
+* `max_freq`                # desired high frequency
+* `epp`                     # energy performance preference
+
+The **core** object attributes which can be adjusted are `max_freq`, `min_freq` and `epp`, min and max set the desired core frequency at which an application wants the core to operate within.
```
Comment (anatolyburakov) [inline pwr.md]: Core?

You already have a doc string above, why duplicate it here? Also, i would reword it something along these lines:
```
Most of the **Core** object attributes are constant and cannot be changed. The only attributes that can be written to by the user, are `min_freq`, `max_freq` and `epp`.
```
The key point we want to get across is that most of the attributes aren't writable.

File: pwr.md [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+import pwr
+
+cores = pwr.get_cores()
+cpus = pwr.get_cpus()
+```
+
+## Adjusting Power Configuration
+### Modifying
+Each cpu and core object have attributes which replicate the capabilities and stats of the physical cores and cpus.
+### cpu
```
Comment (anatolyburakov) [inline pwr.md]: CPU?

File: pwr.md [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+* `all_core_turbo_freq`     # all core turbo frequency
+* `highest_freq`            # highest available frequency
+* `lowest_freq`             # lowest active frequency
+* `uncore_hw_max`           # max available uncore frequency
+* `uncore_hw_min`           # min available uncore frequency
+* `uncore_freq`             # current uncore frequency
+* `uncore_max_freq`         # max desired uncore frequency
+* `uncore_min_freq`         # min desired uncore frequency
+
+### core
```
Comment (anatolyburakov) [inline pwr.md]: Core?

File: pwr.md [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+* `hwp_enabled`             # HWP enabled flag
+* `base_freq`               # base frequency
+* `all_core_turbo_freq`     # all core turbo frequency
+* `highest_freq`            # highest available frequency
+* `lowest_freq`             # lowest active frequency
+* `uncore_hw_max`           # max available uncore frequency
+* `uncore_hw_min`           # min available uncore frequency
+* `uncore_freq`             # current uncore frequency
+* `uncore_max_freq`         # max desired uncore frequency
+* `uncore_min_freq`         # min desired uncore frequency
```
Comment (anatolyburakov) [inline pwr.md]:
```
sst_bf_enabled          # base frequency enabled in BIOS
```
I think you meant SST-BF here.
```
lowest_freq`             # lowest active frequency
```
lowest available?

File: pwr.md [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+* `high_prioirty`           # high/normal priority
+* `base_freq`               # base frequency [2300Mhz]
+* `sst_bf_base_freq`        # priority based frequency [2100Mhz/2700Mhz]
+* `all_core_turbo_freq`     # all core turbo frequency [2800Mhz]
+* `highest_freq`            # highest frequency available [3900Mhz]
+* `lowest_freq`             # lowest active frequency [800Mhz]
+* `curr_freq`               # current core frequency
+* `min_freq`                # desired low frequency
+* `max_freq`                # desired high frequency
+* `epp`                     # energy performance preference
```
Comment (anatolyburakov) [inline pwr.md]:
```
core_id                 # core id number
```
From this it's not clear whether it's physical or logical id of the core.
```
thread_siblings         # list of thread siblings
```
it reads like `frobulate  # frobulate` - this isn't much of an explanation of what "thread siblings" are :) suggested rewording: list of other logical cores residing on the same physical core.
```
high_prioirty           # high/normal priority
```
It isn't immediately obvious what is it, why it's there and when it's used. Suggested rewording:
```
Boolean value indicating whether the core will be set up to be a high priority core when SST-BF is configured.
```

Also, frequencies are SKU-specific, so perhaps a note that the MHz values are for demonstration purposes only?

File: pwr.md [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+
+The **core** object attributes which can be adjusted are `max_freq`, `min_freq` and `epp`, min and max set the desired core frequency at which an application wants the core to operate within. 
+>Available frequencies are between `lowest_freq` and `highest_freq`, with some defined freqeuncies being `base_freq`/`sst_bf_base_freq`/`all_core_turbo_freq` if configuration allows.
+
+`epp` sets the desired performance profile for the core.
+>Available profiles are `default`, `performance`, `balance_performance`, `balance_power` and `power`.
+
+The **cpu** object attributes which can be adjusted by the application are `uncore_max_freq` and `uncore_min_freq`, these set the desired uncore frequency at which an application wants the uncore to operate within.
```
Comment (anatolyburakov) [inline pwr.md]: CPU?

File: pwr.md [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+* `curr_freq`               # current core frequency
+* `min_freq`                # desired low frequency
+* `max_freq`                # desired high frequency
+* `epp`                     # energy performance preference
+
+The **core** object attributes which can be adjusted are `max_freq`, `min_freq` and `epp`, min and max set the desired core frequency at which an application wants the core to operate within. 
+>Available frequencies are between `lowest_freq` and `highest_freq`, with some defined freqeuncies being `base_freq`/`sst_bf_base_freq`/`all_core_turbo_freq` if configuration allows.
+
+`epp` sets the desired performance profile for the core.
+>Available profiles are `default`, `performance`, `balance_performance`, `balance_power` and `power`.
```
Comment (anatolyburakov) [inline pwr.md]: Available profiles depend on the kernel. Do we provide an externally visible list of these? If not, then i suppose documenting these here is OK as it's consistent with what i've seen so far on all the SKU's.

File: pwr.md [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+The **cpu** object attributes which can be adjusted by the application are `uncore_max_freq` and `uncore_min_freq`, these set the desired uncore frequency at which an application wants the uncore to operate within.
+>Available frequencies are between `uncore_hw_max` and `uncore_hw_min`.
+
+```python
+core.min_freq = core.lowest_freq
+core.max_freq = core.highest_freq
+
+cpu.uncore_max_freq = cpu.uncore_hw_max
+cpu.uncore_min_freq = cpu.uncore_hw_min
+```
```
Comment (anatolyburakov) [inline pwr.md]: Maybe add comments explaining what the code does?

File: pwr.md [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+
+### Committing 
+Modification of the power settings of a system is done by altering the core or CPU characteristics, as shown above, 
+then finalizing with the `commit()` function call, this will commit your changes and write to relevant MSRs or sysfs entries.
```
Comment (anatolyburakov) [inline pwr.md]: Where it's written (to sysfs or MSR's) is internal detail, no need to mention this.

File: pwr.md [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+```python 
+cores = pwr.get_cores()
+
+for core in cores:
+    core.min_freq = core.base_freq
+    core.max_freq = core.highest_freq
+    core.commit()
+```
```
Comment (anatolyburakov) [inline pwr.md]: Again, code could do with more comments.

File: pwr.md (hunk 9 above)
Comment (anatolyburakov) [reply pwr.md]: Oh, you _do_ have a note, but it's way below this :)

Perhaps a way around this would be to suppose an imaginary CPU with certain values, and go off that:
```
Given a CPU with min freq at 800, base freq at 2300, max freq at 3900, the profiles will do the following...
```

File: pwr.md [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+* `sst_bf_low_turbo`:   Set high priority cores minimum to 2700Mhz and maximum to 2700Mhz, normal priority cores minimum to 800Mhz and maximum to 3900Mhz.
+>Specific frequencies will depend on system and configuration.
+
+```python
+For c in cores:
+    c.commit("sst_bf_base") # Configure cores with the SST_BF configuration
+```
+
+## Refreshing cpu stats
+On library initialization the systems capabilities and stats are retrieved.
```
Comment (anatolyburakov) [inline pwr.md]: No need to mention _when_ capabilities and stats are retrieved - the user only needs to know that they will be available in either Core or CPU objects.

File: pwr.md [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+CPU stats can become out of date, such as `curr_freq` or `sst_bf_configured`. These can be refreshed in with `refresh_stats()`.
+It is *advised* that CPU and core objects be updated with `refresh_stats()` before requesting object data.
+
+core.refresh_stats() will update:
+* `curr_freq`
```
Comment (anatolyburakov) [inline pwr.md]: cur?

File: pwr.md [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+* `uncore_freq`
+* `uncore_max_freq` 
+* `uncore_min_freq`
+* `sst_bf_configured`
+
+```python
+for c in cores:
+    c.refresh_stats()
+    if c.sst_bf_configured:
+        c.commit("sst_bf_high_turbo")
```
Comment (anatolyburakov) [inline pwr.md]: This seems a little circular. Perhaps you meant `if not c.sst_bf_configured:`? Otherwise you're configuring SST-BF only if it already was configured...

File: pwr.md [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+* `default`:            Set all cores minimum to 800Mhz and maximum to 3900Mhz.
+* `no_turbo`:           Set all cores minimum to 800Mhz and maximum to 2300Mhz.
+* `sst_bf_base`:        Set high priority cores minimum and maximum to 2700Mhz, normal priority cores minimum and maximum to 2100Mhz.
+* `sst_bf_high_turbo`:  Set high priority cores minimum to 2700Mhz and maximum to 3900Mhz, normal priority cores minimum to 800Mhz and maximum to 2100Mhz.
+* `sst_bf_low_turbo`:   Set high priority cores minimum to 2700Mhz and maximum to 2700Mhz, normal priority cores minimum to 800Mhz and maximum to 3900Mhz.
+>Specific frequencies will depend on system and configuration.
+
+```python
+For c in cores:
+    c.commit("sst_bf_base") # Configure cores with the SST_BF configuration
```
Comment (anatolyburakov) [inline pwr.md]: SST-BF

File: pwr.md (hunk 12 above)
Comment (lee-daly) [reply pwr.md]: Where is the duplication?
I would say we need to give the user more info than just "here are the values".

File: pwr.md (hunk 8 above)
Comment (lee-daly) [reply pwr.md]: I don't have this changed to cur in the code, 
is it close to your heart?

File: pwr.md (hunk 24 above)
Comment (lee-daly) [reply pwr.md]: no that should be "sst_bf_enabled"

File: pwr.md (hunk 5 above)
Comment (lee-daly) [reply pwr.md]: from what I can tell, that is what is line is saying.

Comment (anatolyburakov) [reply pwr.md]: Yes, among 10 other points it makes that aren't pertinent :)

File: pwr.md (hunk 8 above)
Comment (anatolyburakov) [reply pwr.md]: Ah, if it's as it is in the code, then OK. I just remember pointing it out on internal review and thought that it was changed. If it wasn't, that's OK.

File: pwr.md (hunk 12 above)
Comment (anatolyburakov) [reply pwr.md]: "min and max set the desired core frequency at which an application wants the core to operate within" is already touched upon by the doc strings. We could perhaps expand on this a bit and provide some general overview of the concepts involved in using this library (such as how to use min/max freqs, epp, etc.), but here is not the place to do it, it should be in a separate section.

File: pwr.md (hunk 8 above)
Comment (anatolyburakov) [reply pwr.md]: disregard :)

File: pwr.md (hunk 23 above)
Comment (anatolyburakov) [reply pwr.md]: disregard :)

File: pwr.md (hunk 24 above)
Comment (anatolyburakov) [reply pwr.md]: If you wanted to demonstrate core stats changing after a profile commit, perhaps you might want to put `refresh_stats` _after_ commit, not (just) before?

File: pwr.md (hunk 8 above)
Comment (lee-daly) [reply pwr.md]: Yes I actually had changed it, but the change must have got lost at some point.

File: pwr.md (hunk 12 above)
Comment (lee-daly) [reply pwr.md]: when you say general overview of the concepts, what else would you say?
I did originally was going to make another section but there wasn't much to it so I put it here.

File: pwr.md (hunk 24 above)
Comment (lee-daly) [reply pwr.md]: no I wasn't demonstrating "core stats changing after a profile commit", I was demonstrating that an application will need to refresh its stats before requesting data.

Rather than before _and_ after any changes.

File: pwr.md (hunk 12 above)
Comment (anatolyburakov) [reply pwr.md]: For example, something like this:

# Give a general overview - e.g. the machine may have one or more CPU's (corresponding to sockets), each CPU has multiple physical cores, and each physical core may optionally have multiple threads (multiple logical cores sitting on the same physical core)
# Explain the CPU part - e.g. CPU has uncore which can be scaled within the limits set by min/max, and it is the place to go when one needs to figure out power consumption (btw this is missing from your API doc - i forgot to add it to readme, apologies for that!)
# Explain the core part - e.g. core can scale its frequency up or down, but only if all thread siblings agree on min/max frequencies (explain how it works), cores can also use SST-BF to set up their frequencies according to SST-BF configuration, and cores can take advantage of SST-CP by setting a specific EPP policy.

Note that none of the above requires referencing specific object attributes - these are general concepts. I think that's plenty to write about to make it an entire section :)

**Comment by lee-daly** [comment]
@anatolyburakov

File: pwr.md [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+cpu.refresh_stats() will update:
+* `uncore_freq`
+* `uncore_max_freq` 
+* `uncore_min_freq`
+* `sst_bf_configured`
+
+```python
+for c in cores:
+    c.refresh_stats()
+    if not c.sst_bf_enabled:
```
Comment (anatolyburakov) [inline pwr.md]: still not sure of this code. if SST-BF is not enabled, why would you commit SST-BF-related profile? i think you meant configured?

File: pwr.md [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+```python
+for cpu in cpus:
+    if cpu.core_list[0].epp == "performance":  # Access core performance profile through core list of CPU object
+        cpu.uncore_max_freq = cpu.uncore_hw_max  # set desired max uncore frequency from hardware limit
+        cpu.uncore_min_freq = cpu.uncore_hw_min  # set desired min uncore frequency from hardware limit
+        cpu.commit() #Commit changes to hardware
+```
+
+## Power Consumption
+The power consumption of the CPU can be read from the `power_consumption` attribute, this value can be compared against the `tdp` value to check is the current power draw close to the limit, indicated by the tdp value.
```
Comment (anatolyburakov) [inline pwr.md]: i would add the following:
```
The power consumption is reported as average since last time it was calculated. The first time this value is read, it will be 0. The next time it is read, it will show average power consumption (in Watts) since last read. Time period between reads must not exceed 60 seconds, otherwise the value will be reset to 0.
```

File: pwr.md (hunk 26 above)
Comment (lee-daly) [reply pwr.md]: yes, mix-up.

File: pwr.md (hunk 27 above)
Comment (lee-daly) [reply pwr.md]: Does this mean the user will have to read it twice the first time? or is that initial read done on initialization? 
or is it more, the use case of this is to set a period of which you would like to measure power consumption i.e set a beginning and end point, rather than just getting the consumption at that very point.

this is turning into a rant, but I assume we cannot get a value for power consumption at a single instance, just a relative one?

Comment (anatolyburakov) [reply pwr.md]: the initial read _is_ done on initialization, but in the general case, yes, if you want to measure power consumption over a period of time, you need to measure twice - that's kind of the definition of an "average" :)

there is no way to read "current power consumption" from the CPU - it is only possible to read how much power it has consumed thus far.

Comment (lee-daly) [reply pwr.md]: Got it Thanks!

Comment (anatolyburakov) [reply pwr.md]: actually, i'm now not so sure it is done on init. Do we call refresh_stats() on init, or do we initialize values internally, without specifically calling the function?

Comment (lee-daly) [reply pwr.md]: no we call refresh_stats, which is better i guess.

Comment (anatolyburakov) [reply pwr.md]: then it's done on init. (and it's better!)

Comment (lee-daly) [reply pwr.md]: do we still want to say "The first time this value is read, it will be 0.", since we dont know this for sure,

Comment (anatolyburakov) [reply pwr.md]: _can_ be 0? it would depend on when it was read. i.e. if you initialize the library but only do the first refresh a minute later?

Comment (lee-daly) [reply pwr.md]: exactly, okay I'll word smith it. :)

File: pwr.md [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+```python
+for cpu in cpus:
+    if cpu.core_list[0].epp == "performance":  # Access core performance profile through core list of CPU object
+        cpu.uncore_max_freq = cpu.uncore_hw_max  # set desired max uncore frequency from hardware limit
+        cpu.uncore_min_freq = cpu.uncore_hw_min  # set desired min uncore frequency from hardware limit
+        cpu.commit() #Commit changes to hardware
+```
+
+## Power Consumption
+The power consumption of the CPU can be read from the `power_consumption` attribute, this value can be compared against the `tdp` value to check is the current power draw close to the limit, indicated by the tdp value. The power consumption is reported as average since last time it was calculated. The first time this value is read, it can be 0. The next time it is read, it will show average power consumption (in Watts) since last read. Time period between reads must not exceed 60 seconds, otherwise the value will be reset to 0, hence, why the initial read may be 0, depending on time between initilization and the first read.
```
Comment (anatolyburakov) [inline pwr.md]: I don't think you need to mention why it may or may not be 0. Just mention that it _can_, and that the user is expected to discard the first read in a sequence of periodic reads.

File: pwr.md [hunk 29]
```diff
@@ -0,0 +1,147 @@
...
+```
+
+## Power Consumption
+The power consumption of the CPU can be read from the `power_consumption` attribute, this value can be compared against the `tdp` value to check is the current power draw close to the limit, indicated by the tdp value. The power consumption is reported as average since last time it was calculated. The first time this value is read, it can be 0. The next time it is read, it will show average power consumption (in Watts) since last read. Time period between reads must not exceed 60 seconds, otherwise the value will be reset to 0, hence, why the initial read may be 0, depending on time between initilization and the first read. 
+
+## Concept Overview
+### CPU
+A CPU may have multiple physical cores, which are represented by the core objects. These cores can optionally have multiple threads, which means there would be multiple logical cores on a single physical core, these are called thread siblings. Each logical core will also have its own core object. Uncore frequency is the frequency at which everything on the CPU, except the cores, runs at. This can be scaled within the limits of min and max. 
+### Core
```
Comment (anatolyburakov) [inline pwr.md]: I don't see us mentioning possible EPP values anywhere? It was there in the first version, but got lost somewhere along the way.

Comment (anatolyburakov) [reply pwr.md]: I think the best place to put it would be alongside the list of profiles.

File: pwr.md [hunk 30]
```diff
@@ -0,0 +1,147 @@
...
+    if cpu.core_list[0].epp == "performance":  # Access core performance profile through core list of CPU object
+        cpu.uncore_max_freq = cpu.uncore_hw_max  # set desired max uncore frequency from hardware limit
+        cpu.uncore_min_freq = cpu.uncore_hw_min  # set desired min uncore frequency from hardware limit
+        cpu.commit() #Commit changes to hardware
+```
+
+## Power Consumption
+The power consumption of the CPU can be read from the `power_consumption` attribute, this value can be compared against the `tdp` value to check is the current power draw close to the limit, indicated by the tdp value. The power consumption is reported as average since last time it was calculated. The first time this value is read, it can be 0. The next time it is read, it will show average power consumption (in Watts) since last read. Time period between reads must not exceed 60 seconds, otherwise the value will be reset to 0, hence, why the initial read may be 0, depending on time between initilization and the first read. 
+
+## Concept Overview
```
Comment (anatolyburakov) [inline pwr.md]: Now that i am looking at the file in its entirety, i think this section is out of place at the end. It is better to put it somewhere in the beginning, before the API explanations.

File: pwr.md [hunk 31]
```diff
@@ -0,0 +1,154 @@
...
+        cpu.commit() #Commit changes to hardware
+```
+
+## EPP
+The EPP value (`epp` attribute in the `Core` object) uses SST-CP technology to prioritize core power consumption. Available EPP values are:
+* performance (maximum power consumption priority)
+* balance_performance (default power consumption priority)
+* balance_power (lower power consumption priority)
+* power (lowest power consumption priority)
+for more information about EPP, see relevant product manuals' section describing the SST-CP technology
```
Comment (anatolyburakov) [inline pwr.md]: Newline before this, and starting from capital, and a dot at the end :)
